openwisp_utils/admin_theme/admin.py

```diff
@@ -18,6 +27,70 @@ class OpenwispAdminSite(admin.AdminSite):
     )
     enable_nav_sidebar = False
 
+    def login(self, *args, **kwargs):
+        response = super().login(*args, **kwargs)
+        if isinstance(response, HttpResponseRedirect):
+            response = HttpResponseRedirect(reverse('admin:ow_dashboard'))
+        return response
+
+    def ow_dashboard(self, request, extra_context=None):
+        context = {
+            'is_popup': False,
+            'has_permission': True,
+        }
+        schema = copy.deepcopy(DASHBOARD_SCHEMA)
+
+        for key, value in schema.items():
+            app_label, model_name, group_by = value['query_params'].values()
+            try:
+                model = load_model(app_label, model_name)
+            except ImproperlyConfigured:
+                raise ImproperlyConfigured(
+                    f'Error adding dashboard element {key}.'
+                    f'REASON: {app_label}.{model_name} could not be loaded.'
+                )
+
+            # Filter query according to organization of user
+            if hasattr(model, 'organization_id') and not request.user.is_superuser:
+                qs = model.objects.filter(
+                    organization_id__in=request.user.organizations_managed
+                )
```

Comment:
In the case of the monitoring module, the query will be something like:

```python
DeviceMonitoring.objects.filter(device__organization_pk__in=request.user.organizations_managed)
```

So this part needs to be improved to allow modules to supply the information on where to look for the organization field.

One possibility is to add another argument to `query_params`:

- `organization_lookup`, which defaults to `organization__in` but can be overridden, eg:

```python
register_dashboard_element(
    'Device Health Status',
    {
        'query_params': {
            'app_label': 'device_monitoring',
            'model': 'DeviceMonitoring',
            'group_by': 'status',
            'organization_lookup': 'device__organization__in',
        },
        'colors': {
            'unknown': 'gray'
            'ok': 'green',
            'problem': 'yellow',
            'critical': 'red',
        },
    },
)
```

The code can then become something like:

```python
organization_lookup = value['query_params'].pop('organization_lookup', 'organization__in')
app_label, model_name, group_by = value['query_params'].values()
qs = model.objects.filter(**{organization_lookup: request.user.organizations_managed})
```

Reply:
@nemesisdesign  in openwisp_monitoring I was thinking to write define element configuration as below:

```python
register_dashboard_element(
    'Monitoring Status',
    {
        'query_params': {
            'app_label': 'config',
            'model': 'device',
            'group_by': 'monitoring__status',
        },
        'colors': {
            'unknown': 'grey',
            'ok': 'green',
            'critical': 'orange',
            'problem': 'red',
        },
    },
) 
```

Similarly for openwisp-controller it would have become:

```python
register_dashboard_element(
    'Configuration Status',
    {
        'query_params': {
            'app_label': 'config',
            'model': 'device',
            'group_by': 'config__status',
        },
        'colors': {'applied': 'green', 'modified': 'orange', 'error': 'red'},
    },
)
```
-----------

I know it will create a join in the database which will increase the time of execution of the query, but I wanted to keep it simple. The admin class of `Device` in both openwisp-controller and openwisp-monitoring has already implemented filters on based of `config__status` and `monitoring__status` respectively. If we use `Config` or `DeviceMonitoring` models in the element configuration, we will have to pass other arguments as well like where to redirect while clicking on a section of pie chart. 

The current implementation will work with above code. 

------------------

I have opened PRs in openwisp-controller and openwisp-monitoring for adding elements to the dashabord:

- https://github.com/openwisp/openwisp-controller/pull/369
- https://github.com/openwisp/openwisp-monitoring/pull/264

Reply:
I left comments in each PR.